redash/schema.py

```diff
@@ -0,0 +1,186 @@
+
+    # Find the columns that need to be created by subtracting the sets:
+    persisted_column_set = set([col_data["name"] for col_data in persisted_column_data])
+    columns_to_create = existing_columns_set.difference(persisted_column_set)
+
+    column_metadata = [column_data[col_name] for col_name in columns_to_create]
+
+    models.db.session.bulk_insert_mappings(ColumnMetadata, column_metadata)
+
+
+class SchemaCache(object):
+    """
+    This caches schema requests in redis and uses a method to
+    serve stale values while the cache is being populated or
+    updated to handle the thundering herd problem.
+    """
+
+    timeout = settings.SCHEMAS_REFRESH_SCHEDULE * 60
+    # keeping the stale cached items for 10 minutes longer
+    # than its timeout to make sure repopulation can work
+    stale_cache_timeout = 60 * 10
+
+    def __init__(self, data_source):
+        self.data_source = data_source
+        self.client = redis_connection
+        self.cache_key = "data_source:schema:cache:{}".format(self.data_source.id)
+        self.lock_key = "{}:lock".format(self.cache_key)
+        self.fresh_key = "{}:fresh".format(self.cache_key)
+
+    def load(self):
+        """
+        When called will fetch all table and column metadata from
+        the database and serialize it with the TableMetadataSerializer.
+        """
+        # due to the unfortunate import time side effects of
+        # Redash's package layout this needs to be done inline
+        from redash.serializers import TableMetadataSerializer
+
+        schema = []
+        tables = (
+            TableMetadata.query.filter(
+                TableMetadata.data_source_id == self.data_source.id,
+                TableMetadata.exists.is_(True),
+            )
+            .order_by(TableMetadata.name)
+            .options(
+                joinedload(TableMetadata.columns),
+                joinedload(TableMetadata.sample_queries),
+            )
+        )
+
+        for table in tables:
+            schema.append(
+                TableMetadataSerializer(table, with_favorite_state=False).serialize()
+            )
+        return schema
+
+    def get_schema(self, refresh=False):
+        """
+        Get or set the schema from Redis.
+
+        This will first check for the fresh key and either
+        return the schema value if it's still fresh or
+        repopulate the cache key and return the stale value.
+
+        This will refresh the schema from the data source's API
+        when requested with the refresh parameter, which will also
+        (re)populate the cache.
+        """
+        if refresh:
+            from redash.tasks.queries import refresh_schema
+
+            refresh_schema.apply_async(
+                args=(self.data_source.id,), queue=settings.SCHEMAS_REFRESH_QUEUE,
+            )
+        # First check if the fresh key is still there
+        is_fresh = redis_connection.get(self.fresh_key)
```

Comment:
Is this maybe a bit misleading? It says we are first checking that the fresh key is still there, but that doesn't happen until `if is_fresh` below right?

Reply:
You're right, fixed in https://github.com/mozilla/redash/pull/995/commits/d7b12cb69227f03c98113355e1049693103ce647.